homeassistant/components/device_tracker/__init__.py

```diff
@@ -236,10 +238,51 @@ def see(hass: HomeAssistantType, mac: str = None, dev_id: str = None,
 
 async def async_setup_entry(hass, entry):
     """Set up an entry."""
-    await hass.data[DOMAIN](entry.domain, entry)
+    await hass.data[DOMAIN].get_async_setup_platform()(entry.domain, entry)
     return True
 
 
+async def async_unload_entry(hass, entry):
+    """Unload an entry."""
+    platform = hass.data[DOMAIN].get_platform(entry.domain)
+    if platform and hasattr(platform, 'async_unload_entry'):
+        await platform.async_unload_entry(hass, entry)
+        return True
+    return False
+
+
+class PlatformManager:
+    """Store data needed to unload entries from loaded platforms."""
+
+    def __init__(self, async_setup_platform: callable):
+        """Initialize a platform manager."""
+        self.platforms = {}
+        self.async_setup_platform = async_setup_platform
+
+    def add_platform(self, platform_name: str, platform):
+        """Add a platform that has been loaded to the platform manager."""
+        _LOGGER.debug('Adding %s to PlatformManager', platform_name)
+        if platform_name is None:
+            return
+        if platform_name in self.platforms:
+            _LOGGER.warning(
+                'Cannot add duplicate platform %s to platform manager',
+                platform_name
+            )
+            return
+        self.platforms[platform_name] = platform
+
+    def get_platform(self, platform_name: str):
+        """Fetch a platform that has been loaded from the platform manager."""
+        if platform_name in self.platforms:
+            return self.platforms[platform_name]
+        return None
+
+    def get_async_setup_platform(self):
```

Comment:
We don't need this method. We can just call `async_setup_platform` attribute on the `PlatformManager` instance.

Reply:
Yep, my Java day job leaking in... setters/getters for everything... :)